src/pax_store/Page.h

```diff
@@ -0,0 +1,248 @@
+        }
+        cout << endl;
+    }
+
+    T *readRow(RowIndex index) {
+        T *returnRow = new T[*(this->numberOfAttributes)];
+        for (unsigned i = 0; i < *(this->numberOfAttributes); i++) {
+            Header offset = *(minipageOffsets + i);
+
+            T *minipage = (T *)(start + offset / 2);
+            T *location = minipage + index;
+            returnRow[i] = *location;
+        }
+        return returnRow;
+    }
+
+    T *projectRow(RowIndex index, std::vector<uint64_t> &projection) {
+        T *returnRow = new T[projection.size()];
+        for (unsigned i = 0; i < projection.size(); i++) {
+            Header offset = *(minipageOffsets + projection[i]);
+
+            T *minipage = (T *)(start + offset / 2);
+            T *location = minipage + index;
+            returnRow[i] = *location;
+        }
+        return returnRow;
+    }
+
+    void writeRecord(const T *record) {
+        unsigned short recordSize = *numberOfAttributes * sizeof(T);
+        if (*freeSpace < recordSize) {
+            cout << "Page is full" << endl;
+            throw;
+        }
+
+        // cout << "Write: ";
+        for (int i = 0; i < *numberOfAttributes; i++) {
+            Header offset = *(minipageOffsets + i);
+
+            T *minipage = (T *)(start + offset / 2);
+            T *location = minipage + *numberOfRecords;
+            // cout << record[i] << " (" << location << ") ";
+            *(location) = record[i];
+        };
+        // cout << endl;
+        (*numberOfRecords)++;
+        (*freeSpace) -= recordSize;
+    }
+
+    vector<unsigned> query(std::vector<Filters::Scalar::Filter<T> *> &filters) {
+        vector<unsigned> positions;
+        bool firstRun = true;
+
+        for (auto &filter : filters) {
+            // Find the minipage for the attribute to which the filter applies.
+            Header offset = *(minipageOffsets + filter->index);
+            T *minipage = (T *)((char *)start + offset);
+
+            if (firstRun) {
+                // On the first run, we need to go through all records.
+                for (RowIndex index = 0; index < *numberOfRecords; index++) {
+                    T cell = *(minipage + index);
+                    if (filter->match(cell)) positions.push_back(index);
+                }
+                firstRun = false;
+            } else {
+                // On subsequent runs, we only need to go through the positions where
+                // all previous filters already matched.
+                for (auto &position : positions) {
+                    T cell = *(minipage + position);
+                    if (filter->match(cell)) positions.push_back(position);
+                }
+            }
+        }
+
+        return positions;
+    }
+
+    // Note: This function currently only works when T is set to uint64_t. Otherwise, the `ColumStore::Helper::store` function does not
+    // work.
+    // TODO: Allow for uint32_t values
```

Comment:
Some methods in `ColumnStore::Helper` currently do not work as they are and this has a reason.
@adbexy urged me to use `uint64_t` as indices (positions) for the table to support more than `2^32` rows as he had doubts that we could create tables with enough data for our benchmarks. But since we could have >16GB data in a **single** column using 32 bit data I dont think this argument holds up. Nevertheless I wanted to ask Prof. Habich on his opinion on the matter but our recent meetings got cancelled.

The problem with the 64 bit indices is, that only 8 of them fit into our registers. If we have 32 bit data stored in our table we have a mismatch: 16 values but only 8 indices. This is why the current 64 bit indices only work with 64 bit data.

Reply:
Okay, then we will simply ask Prof. Habich and see what he has to say. But thanks for the explanation 👍🏻